### PR TITLE
AutomountServiceAccountToken property implemented

### PR DESCRIPTION
**Pull request by ph311o**
Added new property "automountServiceAccountToken" in ServiceAccounts and Deployments necessary for restricted environments.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/charts/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] README is updated with new configuration values *(if applicable)*

Fixes #

### CODE REVIEW

File: keda/Chart.yaml
Comment (tomkerkhove) [inline keda/Chart.yaml:11]: Can you revert this please? We typically bump when we do a release

File: keda/templates/12-keda-deployment.yaml
Comment (tomkerkhove) [inline keda/templates/12-keda-deployment.yaml:42]: Let's use this approach instead:
```suggestion
      automountServiceAccountToken: {{ .Values.serviceAccount.automount.name }}
```

File: keda/templates/22-metrics-deployment.yaml
Comment (tomkerkhove) [inline keda/templates/22-metrics-deployment.yaml:45]: Let's use this approach instead:
```suggestion
      automountServiceAccountToken: {{ .Values.serviceAccount.automount.name }}
```

File: keda/values.yaml
Comment (tomkerkhove) [inline keda/values.yaml:51]:
```suggestion
# Specifies that both deployments (KEDA/Metrics) should automount API-Credentials
```

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
Did a quick review, thanks for the change!

@zroubalik Do you want to have these changes in our generated YAML files?

File: keda/Chart.yaml
Comment (ph311o) [reply keda/Chart.yaml]: Done.

File: keda/templates/12-keda-deployment.yaml
Comment (ph311o) [reply keda/templates/12-keda-deployment.yaml:42]: @tomkerkhove Thank you for suggested change. Unfortunately this will not work as autmountServiceAccountToken is a boolean field. Because of security restrictions we need to set autmountServiceAccountToken in serviceAccount to false and in deployments to true.

File: keda/templates/22-metrics-deployment.yaml
Comment (ph311o) [reply keda/templates/22-metrics-deployment.yaml]: @tomkerkhove Thank you for suggested change. Unfortunately this will not work as autmountServiceAccountToken is a boolean field. Because of security restrictions we need to set autmountServiceAccountToken in serviceAccount to false and in deployments to true.

File: keda/values.yaml
Comment (ph311o) [reply keda/values.yaml]: Done

File: keda/templates/12-keda-deployment.yaml
Comment (tomkerkhove) [reply keda/templates/12-keda-deployment.yaml:42]: Let's go with this then?
```suggestion
      automountServiceAccountToken: {{ .Values.serviceAccount.automount.enabled }}
```

File: keda/templates/22-metrics-deployment.yaml
Comment (tomkerkhove) [reply keda/templates/22-metrics-deployment.yaml]: Let's go with this then?
```suggestion
      automountServiceAccountToken: {{ .Values.serviceAccount.automount.enabled }}
```

File: keda/templates/12-keda-deployment.yaml
Comment (tomkerkhove) [reply keda/templates/12-keda-deployment.yaml:42]: Oh I get it. So it's not related to the service account but rather the deployment instead?

What about this then?
```suggestion
      automountServiceAccountToken: {{ .Values.deployment.automount.enabled }}
```

Comment (tomkerkhove) [reply keda/templates/12-keda-deployment.yaml:42]: That's OK for me, thanks! I'm not an expert in this area, does `automountServiceAccountToken` have requirements or involve any risk to break existing installations?

Comment (ph311o) [reply keda/templates/12-keda-deployment.yaml]: I have set serviceAccount and deployments to default behaviour of k8s so there is no risk for breaking existing installations. Additionally the needed opt-in for mounting API-token is added to both keda and metrics deployment. Please review. Thank you for your support.

**Comment by JorTurFer** [comment]
Hey,
I'm looking for information about this feature and I can't find any clear use case (or most probably, I'm not understanding them). 
Could you explain your use case?

**Comment by ph311o** [comment]
@JorTurFer Hi. Our UseCase is as follow:

In general Kubernetes automounts service account credentials which allows any compromised pod to run API commands against the cluster. We are restricted to stop this behaviour by specifying automountServiceAccountToken: false in ServiceAccount. This is a global rule so we cannot whitelist the helm chart.

As Keda and metrics deployments/pods need API access we need to opt in for these two with automountServiceAccountToken: true in both deployment.yamls

When you set both ServiceAccount and Deployments automountServiceAccountToken: true it's the same like not specifying at all so I am not sure if it makes sense to specify it in the installation.yaml. We need it in Helm Chart as said to disable default ServiceAccount behaviour and ensure functionality. 

Also see:

- https://github.com/kubernetes/kubernetes/issues/57601
- https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Grundschutz/IT-GS-Kompendium_Einzel_PDFs_2022/06_APP_Anwendungen/APP_4_4_Kubernetes_Edition_2022.pdf;jsessionid=831EB9783FD3996CD80511A21925E8EE.internet482?__blob=publicationFile&v=3#download=1

**Comment by JorTurFer** [comment]
I got the reasons (I don't understand the document in German but I guess that it's an official regulation). 
My doubt is, how will KEDA work without its service account? I mean, KEDA needs its service account with the proper given permissions or basically KEDA doesn't work. 
I understand that common pods without any interaction with the API don't neet that service account, but KEDA needs it to work, Am I missing something? Are you enabling/injecting it in any other way?

**Comment by ph311o** [comment]
@JorTurFer Do you need more informations from my side or do we wait for more opinions by other maintainers? In general you can disable it in ServiceAccount and enable it in Deployments hardcoded. But then everybody who would use the Keda ServiceAccount perhaps for other deployments could not longer access API. This was the reason why I decided to implement a switch in values.yaml for ServiceAccount.

**Comment by zroubalik** [comment]
I think it is okay to have an optional switch, ie. default to the existing behavior.

**Comment by ph311o** [comment]
@JorTurFer Yes absolutely it keeps current behaviour. Would be great if anybody could merge it for next release. Thank you.

File: keda/templates/22-metrics-deployment.yaml
Comment (tomkerkhove) [inline keda/templates/22-metrics-deployment.yaml:45]: Can you move this to a value with default property `true` instead please?

File: keda/templates/12-keda-deployment.yaml
Comment (tomkerkhove) [inline keda/templates/12-keda-deployment.yaml:42]: Can you move this to a value with default property `true` instead please?

**Review by tomkerkhove — CHANGES_REQUESTED** [review]

File: keda/templates/22-metrics-deployment.yaml
Comment (ph311o) [reply keda/templates/22-metrics-deployment.yaml:45]: Yes I can do that. I am not sure if this makes sense as you need that to set to true for having API Access. If you would set it to false KEDA would not work anymore. Please confirm if I should change it.

File: keda/templates/12-keda-deployment.yaml
Comment (ph311o) [reply keda/templates/12-keda-deployment.yaml:42]: Yes I can do that. I am not sure if this makes sense as you need that to set to true for having API Access. If you would set it to false KEDA would not work anymore. Please confirm if I should change it.

Comment (tomkerkhove) [reply keda/templates/12-keda-deployment.yaml:42]: With that you mean if `serviceAccount.automountServiceAccountToken` flag is on true? If so, why not re-use that same value then?

Comment (ph311o) [reply keda/templates/12-keda-deployment.yaml:42]: Let me explain:
Every ServiceAccount generates an API Token regardless of what automountServiceAccountToken is set to (true, false or even not set)

Default behaviour when not specified automountServiceAccountToken in ServiceAccount and Deployment -> Pods using this ServiceAccount can access the API-Token.

ServiceAccount.automountServiceAccountToken: false -> All pods using this ServiceAccount have no access to API-Token anymore (but the ServiceAccount still have an API-Token stored internally)

Deployment can opt-in to regain access to API-Token:
Deployment.automountServiceAccountToken: true -> pods regain access to token

Does a switch make sense for ServiceAccount.automountServiceAccountToken?
Yes because e.g. we have a security policy that we can only deploy ServiceAccounts with ServiceAccount.automountServiceAccountToken: false.
So we have to define Deployment.automountServiceAccountToken: true in every Deployment which needs API-Access.

If somebody would use the ServiceAccount for own custom Deployments without specifying Deployment.automountServiceAccountToken they need to set ServiceAccount.automountServiceAccountToken: true. (Which would be the default switch value... So they do not recognize a change)

Does a switch make sense for Deployment.automountServiceAccountToken?
No. Because KEDA needs API-Access. A switch which could possible set Deployment.automountServiceAccountToken: false will break the API-Access for KEDA-Pods. (Regardless of ServiceAccount.automountServiceAccountToken)

Why even specifying Deployment.automountServiceAccountToken?
Because when ServiceAccount.automountServiceAccountToken: false and not specifying Deployment.automountServiceAccountToken the API-Access is gone.

Why not re-use the same value?
Because when setting ServiceAccount.automountServiceAccountToken: false then reusing value would result in Deployment.automountServiceAccountToken: false which prevents API-Access for pods.

Comment (tomkerkhove) [reply keda/templates/12-keda-deployment.yaml:42]: Thanks for explaining!

**Review by tomkerkhove — APPROVED** [review]

**Comment by tomkerkhove** [comment]
Would love to have a sign-off of @zroubalik or @JorTurFer as well though

**Review by JorTurFer — APPROVED** [review]
F**k, I missed this PR between the notifications :( 
If the behavior doesn't change, I don't have any problem to merge it

**Comment by tomkerkhove** [comment]
Thank you @ph311o and sorry for being slow on our end!

**Pull request merged**
